python/pyspark/sql/catalog.py

```diff
@@ -256,27 +258,58 @@ def registerFunction(self, name, f, returnType=StringType()):
         >>> spark.sql("SELECT stringLengthInt('test')").collect()
         [Row(stringLengthInt(test)=4)]
 
+        >>> from pyspark.sql.types import IntegerType
+        >>> from pyspark.sql.functions import udf
+        >>> slen = udf(lambda s: len(s), IntegerType())
+        >>> _ = spark.udf.register("slen", slen)
+        >>> spark.sql("SELECT slen('test')").collect()
+        [Row(slen(test)=4)]
+
         >>> import random
         >>> from pyspark.sql.functions import udf
-        >>> from pyspark.sql.types import IntegerType, StringType
+        >>> from pyspark.sql.types import IntegerType
         >>> random_udf = udf(lambda: random.randint(0, 100), IntegerType()).asNondeterministic()
-        >>> newRandom_udf = spark.catalog.registerFunction("random_udf", random_udf, StringType())
+        >>> newRandom_udf = spark.udf.register("random_udf", random_udf)
         >>> spark.sql("SELECT random_udf()").collect()  # doctest: +SKIP
-        [Row(random_udf()=u'82')]
+        [Row(random_udf()=82)]
         >>> spark.range(1).select(newRandom_udf()).collect()  # doctest: +SKIP
-        [Row(random_udf()=u'62')]
+        [Row(<lambda>()=26)]
+
+        >>> from pyspark.sql.functions import pandas_udf, PandasUDFType
+        >>> @pandas_udf("integer", PandasUDFType.SCALAR)  # doctest: +SKIP
+        ... def add_one(x):
+        ...     return x + 1
+        ...
+        >>> _ = spark.udf.register("add_one", add_one)  # doctest: +SKIP
+        >>> spark.sql("SELECT add_one(id) FROM range(3)").collect()  # doctest: +SKIP
+        [Row(add_one(id)=1), Row(add_one(id)=2), Row(add_one(id)=3)]
         """
 
         # This is to check whether the input function is a wrapped/native UserDefinedFunction
         if hasattr(f, 'asNondeterministic'):
-            udf = UserDefinedFunction(f.func, returnType=returnType, name=name,
-                                      evalType=PythonEvalType.SQL_BATCHED_UDF,
-                                      deterministic=f.deterministic)
+            if f.evalType not in [PythonEvalType.SQL_BATCHED_UDF,
+                                  PythonEvalType.SQL_PANDAS_SCALAR_UDF]:
+                raise ValueError(
+                    "Invalid f: f must be either SQL_BATCHED_UDF or SQL_PANDAS_SCALAR_UDF")
+            if returnType is not None and not isinstance(returnType, DataType):
+                returnType = _parse_datatype_string(returnType)
+            if returnType is not None and returnType != f.returnType:
```

Comment:
cc @cloud-fan @taku-k

Reply:
Why did you cc other guys @gatorsmile?

Reply:
I am not sure which one is better.

Reply:
I think we are trying to avoid to set `returnType` at register time. Current way appearently allows to take `returnType` when they are matched. Also, the suggestion allows resemble the overloading of the Scala version we talked. I would like to get this into 2.3 and push forward.

Could you maybe eblabourate why you are not sure? Let me try to explain it.

Reply:
Since the API already has this parameter `returnType`, we should support it if possible. We need to do our best to avoid issuing the unnecessary exception.

Reply:
I see. So,

> Since the API already has this parameter `returnType`, we should support it if possible. 

I get it. So, we want to disallow to set `returnType` at register time but also leave `returnType` support permissively as possible as we can. I think failing fast (it's not in runtime) is fine if it give a clear message.

Consider this, what we want is both APIs:

```
def register(name, f)  # for UDF

def register(name, f, returnType)  # for Python native functions
```

I think we should only support both cases in a better API design.

What I am worried is JIRAs and questions from users that why

```
def register(name, f, retunType)  # for UDF
```

is not supported because we will apparently allow a single case in this way.

Consider another case of this: if we only have this:

```
def register(name, f)  # for UDF
```

calling like this:

```
register("a", lambda x: x, 1)
```

throw an `TypeError` as below:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: register() takes exactly 2 arguments (3 given)
```

If we should allow the same `returnType` later, we can simply allow it because this case is less stricter.

Reply:
Because you can only set the `returnType` of the `UserDefinedFunction`, I think it is confusing to let users to pass in returnType and require it to be matched with `returnType` of the `UserDefinedFunction`.

I think we can throw an exception when the `returnType` is not None for `UserDefinedFunction` case.

Reply:
The function already has the parameter `returnType`. Thus,  `def register(name, f)  # for UDF` is not right.

Reply:
```
    def register(self, name, f, returnType=None):
        return self.sqlContext.registerFunction(name, f, returnType)
```

`returnType` is a parameter already. `def register(name, f)  # for UDF` is not true.

Now, the issue we are facing is whether we should issue an exception or not when users provide the same returnType as the UDF's.

Reply:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: register() takes exactly 2 arguments (3 given)
```

This error message is still confusing to end users. We need to explicitly explain that `returnType` is not allowed when `f` is a UDF.

Reply:
What we want to support is `def register(name, f) # for UDF` so I am saying let's fix it as if we have `def register(name, f) # for UDF` which prints an error when `returnType` is given as described above.

Reply:
I am not saying we should have the same message. I am trying to persuade you to throw an error in this case.

Reply:
Is it common in our current PySpark impl?

Reply:
I might miss something but I think it's okay to take `returnType` parameter optionally if the value is the same as the udf's.

Reply:
Optional value is okay but I mean it's better to throw an exception. I am not seeing the advantage of supporting this optionally. @ueshin do you think it's better to support this case?

I am less sure of the point of supporting `returnType` with UDF when we are disallowed to change. It causes confusion like we allow it but then if the type is different, we will issue an exception.

Is it more important to allow this corner case than we make the APIs clear as if we have `def register(name, f) # for UDF` alone? We can keep clear about disallowing `returnType` at register time too.

Reply:
I see what you mean. Now I became neutral but slightly on your side.